File: dev/benchmarks/prefill4k_allrows_qmv.cpp
// CPU-only ABI/geometry/ID/scratch/alias policy checks. No Metal headers/backend.
#include "prefill4k_allrows_qmv.hpp"

#include <algorithm>
#include <bit>
#include <cmath>
#include <iostream>
#include <string>
#include <vector>

namespace {
namespace qmv = splash::flash::gathered_i8_qmv;
void require(bool value, const char *reason) {
  if (!value) throw std::runtime_error(reason);
}
template<class Function> void rejects(Function &&function, const char *reason) {
  try { function(); } catch (const std::invalid_argument &) { return; }
  throw std::runtime_error(reason);
}
uint16_t bf16(float value) {
  const uint32_t word = std::bit_cast<uint32_t>(value);
  return uint16_t((word + 0x7fffu + ((word >> 16) & 1u)) >> 16);
}
float number(uint16_t value) { return std::bit_cast<float>(uint32_t(value) << 16); }
float sanitized(uint16_t value, uint32_t &diagnostics) {
  if ((value & 0x7f80u) == 0x7f80u) { diagnostics |= qmv::kInvalidNumerics; return 0.0f; }
  return number(value);
}
void cpuChecks() {
  require(sizeof(FlashGatheredI8QMVParams) == 16 && qmv::kThreads == 128,
      "host/shader ABI or fixed launch differs");
  require(qmv::kPolicy != "private-allrows-full512-signed-i8-f32-late-row-scale-bf16-dots-m16tight-small-m32m64-large-no-target-q4gpu-v1",
      "QMV reduction identity aliases MPP control");
  for (uint32_t rows : {1u, 2u, 3u, 4u, 8u, 15u, 16u}) {
    const auto g = qmv::geometry(rows);
    require(g.inputBytes == uint64_t{rows} * 5120 && g.idBytes == uint64_t{rows} * 80 &&
        g.intermediateBytes == uint64_t{rows} * 12800 &&
        g.expertDownBytes == uint64_t{rows} * 51200 &&
        g.gateColumnGroups == 160 && g.downColumnGroups == 640,
        "canonical tensor extent or launch differs");
    // Synthetic CPU metadata ranges, never dereferenced.
    const qmv::ByteView input{0x10000000, g.inputBytes}, ids{0x20000000, g.idBytes},
        activated{0x30000000, g.intermediateBytes}, down{0x40000000, g.expertDownBytes},
        diag{0x50000000, 4};
    qmv::validateViews(g, input, ids, activated, diag, false);
    qmv::validateViews(g, activated, ids, down, diag, true);
    rejects([&] { qmv::validateViews(g, {input.address, input.bytes - 1}, ids, activated, diag, false); },
        "undersized hidden view admitted");
    rejects([&] { qmv::validateViews(g, activated, ids, {down.address, down.bytes - 1}, diag, true); },
        "undersized expert-down view admitted");
    rejects([&] { qmv::validateViews(g, input, {ids.address + 1, ids.bytes}, activated, diag, false); },
        "misaligned signed I64 IDs admitted");
    rejects([&] { qmv::validateViews(g, input, ids, {input.address + input.bytes - 2, activated.bytes}, diag, false); },
        "partial write/input overlap admitted");
    rejects([&] { qmv::validateViews(g, input, ids, activated, {activated.address, 4}, false); },
        "diagnostic/output alias admitted");
    rejects([&] { qmv::validateViews(g, input, ids, {0, activated.bytes}, diag, false); },
        "missing output pointer admitted");
    rejects([&] { qmv::validateViews(g, input, ids,
        {std::numeric_limits<uintptr_t>::max() - 1, activated.bytes}, diag, false); },
        "overflowing address extent admitted");
    require(!qmv::overlaps(input, {input.address + input.bytes, 2}),
        "adjacent disjoint view rejected");
    std::vector<int64_t> routeIDs(uint64_t{rows} * 10);
    for (uint32_t row = 0; row < rows; ++row)
      for (uint32_t slot = 0; slot < 10; ++slot)
        routeIDs[row * 10 + slot] = (row * 73 + slot * 53) % 512;
    require(qmv::fixtureIDsValid(routeIDs, rows) &&
        qmv::fixtureIDDiagnostics(routeIDs, rows, 0x80000000u) == 0x80000000u,
        "valid canonical I64 ownership rejected or sticky bits lost");
    for (int64_t invalid : {int64_t{-1}, int64_t{512}, int64_t{1} << 32,
                           std::numeric_limits<int64_t>::max()}) {
      auto bad = routeIDs; bad.back() = invalid;
      require(!qmv::fixtureIDsValid(bad, rows) &&
          qmv::fixtureIDDiagnostics(bad, rows, 0x80000000u) == 0x80000001u,
          "malformed I64 original ID diagnostics differ");
    }
    auto duplicate = routeIDs; duplicate[1] = duplicate[0];
    require(qmv::fixtureIDsValid(duplicate, rows) &&
        qmv::fixtureIDDiagnostics(duplicate, rows, 0x80000000u) == 0x80000001u,
        "duplicate finite computations must flagID without flagNumeric");
    require(qmv::fixtureIDDiagnostics(std::span(routeIDs).first(routeIDs.size() - 1), rows) == 2,
        "malformed fixture ID extent admitted");
  }
  for (uint32_t rows : {0u, 17u, 256u, UINT32_MAX})
    rejects([&] { (void)qmv::geometry(rows); }, "unsupported row count admitted");
  for (uint32_t selections : {0u, 1u, 9u, 11u, UINT32_MAX})
    rejects([&] { (void)qmv::geometry(1, selections); }, "nonK10 gathered policy admitted");
  require(!qmv::validGeometry(1, 10, 511), "partial inventory admitted");
  std::array<uint32_t, 512> ranks{};
  for (uint32_t expert = 0; expert < 512; ++expert) ranks[expert] = (expert * 173 + 31) % 512;
  auto sorted = ranks; std::sort(sorted.begin(), sorted.end());
  for (uint32_t i = 0; i < 512; ++i) require(sorted[i] == i, "nonidentity fixture ranks not bijective");
  for (uint32_t expert = 0; expert < 512; ++expert) {
    uint32_t sticky = 0x80000000u;
    require(qmv::checkedRank(expert, ranks, sticky) == ranks[expert] && sticky == 0x80000000u,
        "originalID mistakenly used as persisted rank");
  }
  for (uint32_t invalid : {512u, UINT32_MAX}) {
    auto bad = ranks; bad[71] = invalid;
    uint32_t sticky = 0x80000000u;
    require(qmv::checkedRank(71, bad, sticky) == UINT32_MAX && sticky == 0x80000001u,
        "Full512 missing/corrupt rank must flagID");
  }
  require(bf16(-0.0f) == 0x8000 && bf16(1.00390625f) == 0x3f80,
      "BF16 fixture rounding or signed zero differs");
  uint32_t sticky = 0x80000000u;
  require(sanitized(0x7fc0, sticky) == 0 && sanitized(0x7f80, sticky) == 0 &&
      sanitized(0xff80, sticky) == 0 && sticky == 0x80000004u,
      "original nonfinite zero/sticky behavior differs");
  require(std::signbit(sanitized(0x8000, sticky)), "finite signed zero changed");
  // Proves the scale boundary is meaningful. Early per-code BF16 rounding is
  // not the candidate's late-scale signed-I8 dot policy.
  const float late = (float(int8_t{127}) + float(int8_t{-126})) * 0.003f;
  const float early = number(bf16(float(int8_t{127}) * 0.003f)) +
      number(bf16(float(int8_t{-126}) * 0.003f));
  require(bf16(late) != bf16(early), "late-scale adversarial fixture lacks discrimination");
  const char *old = std::getenv(qmv::kFlag);
  const std::string saved = old ? old : "";
  const bool had = old != nullptr;
  ::unsetenv(qmv::kFlag); require(!qmv::requested(), "unset flag must select MPP control");
  ::setenv(qmv::kFlag, "0", 1); require(!qmv::requested(), "flag0 must select MPP control");
  ::setenv(qmv::kFlag, "1", 1); require(qmv::requested(), "flag1 must select QMV");
  for (const char *bad : {"", "true", "01", " 1", "2", "-1"}) {
    ::setenv(qmv::kFlag, bad, 1);
    rejects([] { (void)qmv::requested(); }, "malformed runtime policy flag admitted");
  }
  if (had) ::setenv(qmv::kFlag, saved.c_str(), 1); else ::unsetenv(qmv::kFlag);
}
} // namespace
int main(int argc, char **argv) {
  try {
    if (argc != 2 || std::string_view(argv[1]) != "--cpu-self-test")
      throw std::invalid_argument("CPU-only helper requires --cpu-self-test; no GPU mode exists");
    cpuChecks();
    std::cout << "{\"cpu_checks\":\"passed\",\"gpu_work\":false,\"numerical_qualification\":\"pending\",\"policy\":\""
              << qmv::kPolicy << "\"}\n";
    return 0;
  } catch (const std::exception &error) { std::cerr << error.what() << '\n'; return 1; }
}

File: dev/benchmarks/prefill4k_allrows_qmv.py
#!/usr/bin/env python3
"""Private Store-only gathered-I8 QMV transform; import never performs inference.

Compose after prefill4k_allrows_store.transform in a NEW copied source tree.
Executor routing is the composing parent's responsibility. This does not change
production or the existing MPP methods. Flag0 is the original MPP control.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
FLAG = "SPLASH_FLASH_ALLROWS_I8_GATHERED_QMV"
POLICY = "private-gathered-signed-i8-bf16-input-f32-lane32-strided-dot-late-row-scale-bf16-dots-swiglu-sg4-c1-rows1to16-v1"
HEADER_RELATIVE = "runtime/flash/FlashGatheredI8QMV.hpp"
METAL_RELATIVE = "runtime/metal/kernels/shared/flash_gathered_i8_qmv.metal"
MARKER = "// Private gathered I8 QMV Store overlay v1."


def replace(text: str, before: str, after: str, count: int = 1) -> str:
    actual = text.count(before)
    if actual != count:
        raise RuntimeError(f"Private gathered QMV source drift: expected{count}, got{actual}: {before!r}")
    return text.replace(before, after)


def extra_files() -> dict[str, str]:
    """Content to stage before compiling; no mutation occurs here."""
    base = ROOT / "dev/benchmarks/prefill4k_allrows_qmv"
    return {
        HEADER_RELATIVE: base.with_suffix(".hpp").read_text(),
        METAL_RELATIVE: base.with_suffix(".metal").read_text(),
    }


def transform(relative: str, text: str) -> str:
    if relative in ("runtime/flash/FlashInt8ExpertStore.hpp", "runtime/flash/FlashInt8ExpertStore.mm"):
        if MARKER in text:
            raise RuntimeError("Private gathered I8 QMV transform already applied")
        text = MARKER + "\n" + text
    if relative == "runtime/flash/FlashInt8ExpertStore.hpp":
        text = replace(text, '#include "FlashMoEBlocked.hpp"', '#include "FlashMoEBlocked.hpp"\n#include "FlashGatheredI8QMV.hpp"')
        text = replace(text, "  uint64_t large_row_full_inventory_graph_calls = 0;", """  uint64_t large_row_full_inventory_graph_calls = 0;
  // Graph construction, not GPU completion. QMV only physical rows1..16.
  uint64_t gathered_qmv_gate_up_graph_calls = 0, gathered_qmv_gate_up_graph_rows = 0;
  uint64_t gathered_qmv_down_graph_calls = 0, gathered_qmv_down_graph_rows = 0;""")
        return replace(text, "private:\n  struct Impl;", DECLARATIONS + "private:\n  struct Impl;")
    if relative != "runtime/flash/FlashInt8ExpertStore.mm":
        return text
    text = replace(text, "  std::string numericalIdentity;", """  std::string numericalIdentity;
  const bool gatheredQMV = gathered_i8_qmv::requested();
  mutable std::atomic<uint64_t> qmvGateCalls{0}, qmvGateRows{0}, qmvDownCalls{0}, qmvDownRows{0};""")
    text = replace(text, '    const std::string derivative = std::string("splash.private-allrows-target-v1\\nsource=") +', '    std::string derivative = std::string("splash.private-allrows-target-v1\\nsource=") +')
    text = replace(text, "    numericalIdentity = hash(derivative.data(), derivative.size());", """    if (gatheredQMV)
      derivative += std::string("small_row_policy=") + std::string(gathered_i8_qmv::kPolicy) + "\\n";
    numericalIdentity = hash(derivative.data(), derivative.size());""")
    text = replace(text, "      impl_->largeFullCalls.load(std::memory_order_relaxed)};", """      impl_->largeFullCalls.load(std::memory_order_relaxed),
      impl_->qmvGateCalls.load(std::memory_order_relaxed), impl_->qmvGateRows.load(std::memory_order_relaxed),
      impl_->qmvDownCalls.load(std::memory_order_relaxed), impl_->qmvDownRows.load(std::memory_order_relaxed)};""")
    return replace(text, "} // namespace splash::flash", METHODS + "} // namespace splash::flash")


DECLARATIONS = r'''
  // Store methods keep immutable base mappings and rank allocations private.
  // Dispatch-bound MetalBuffer copies retain their bases until graph disposal.
  [[nodiscard]] bool gatheredQMVEnabled() const;
  void addGatheredQMVGateUp(metal::CommandGraph &graph, uint32_t layer,
      metal::MetalBuffer input, metal::MetalBuffer originalExpertIDs,
      metal::MetalBuffer canonicalIntermediate, metal::MetalBuffer diagnostics,
      uint32_t rows, uint32_t selections = 10) const;
  void addGatheredQMVDown(metal::CommandGraph &graph, uint32_t layer,
      metal::MetalBuffer canonicalIntermediate, metal::MetalBuffer originalExpertIDs,
      metal::MetalBuffer canonicalExpertDown, metal::MetalBuffer diagnostics,
      uint32_t rows, uint32_t selections = 10) const;
'''

METHODS = r'''
namespace {
void gatheredQMVViews(const gathered_i8_qmv::Geometry &g,
    metal::MetalBuffer input, metal::MetalBuffer ids, metal::MetalBuffer output,
    metal::MetalBuffer diagnostics, bool down) {
  requireBytes(input, down ? g.intermediateBytes : g.inputBytes);
  requireBytes(ids, g.idBytes);
  requireBytes(output, down ? g.expertDownBytes : g.intermediateBytes);
  requireBytes(diagnostics, 4);
  const auto view = [](const metal::MetalBuffer &b) {
    return gathered_i8_qmv::ByteView{reinterpret_cast<uintptr_t>(b.contents()), b.sizeBytes()};
  };
  gathered_i8_qmv::validateViews(g, view(input), view(ids), view(output), view(diagnostics), down);
}
} // namespace
bool FlashInt8ExpertStore::gatheredQMVEnabled() const {
  if (!impl_) fail("private gathered I8 QMV Store was moved or disposed");
  if (gathered_i8_qmv::requested() != impl_->gatheredQMV)
    fail("private gathered I8 QMV flag changed after Store construction");
  return impl_->gatheredQMV;
}
void FlashInt8ExpertStore::addGatheredQMVGateUp(metal::CommandGraph &graph, uint32_t index,
    metal::MetalBuffer input, metal::MetalBuffer originalExpertIDs,
    metal::MetalBuffer canonicalIntermediate, metal::MetalBuffer diagnostics,
    uint32_t rows, uint32_t selections) const {
  if (!gatheredQMVEnabled()) fail("private gathered I8 QMV gate/up requires frozen flag1");
  const auto g = gathered_i8_qmv::geometry(rows, selections);
  const auto &layer = impl_->layer(index);
  if (impl_->metadata.layers[index].selectedIDs.size() != 512)
    fail("private gathered I8 QMV requires Full512");
  gatheredQMVViews(g, input, originalExpertIDs, canonicalIntermediate, diagnostics, false);
  for (const auto &buffer : {input, originalExpertIDs, canonicalIntermediate, diagnostics})
    impl_->immutableDisjoint(buffer);
  requireBytes(layer.ranks, 512 * 4);
  for (uint32_t p = 0; p < 2; ++p) {
    requireBytes(layer.codes[p], uint64_t{512} * 640 * 2560);
    requireBytes(layer.scales[p], uint64_t{512} * 640 * 4);
  }
  graph.add("flash_gathered_i8_qmv_gate_up_sg4_c1", {input, layer.codes[0], layer.scales[0],
      layer.codes[1], layer.scales[1], layer.ranks, originalExpertIDs, canonicalIntermediate,
      diagnostics}, FlashGatheredI8QMVParams{rows, selections, 512, 0},
      {g.gateColumnGroups, rows, selections}, {gathered_i8_qmv::kThreads, 1, 1});
  impl_->recordGraph(false, rows, 512);
  impl_->qmvGateCalls.fetch_add(1, std::memory_order_relaxed);
  impl_->qmvGateRows.fetch_add(rows, std::memory_order_relaxed);
}
void FlashInt8ExpertStore::addGatheredQMVDown(metal::CommandGraph &graph, uint32_t index,
    metal::MetalBuffer canonicalIntermediate, metal::MetalBuffer originalExpertIDs,
    metal::MetalBuffer canonicalExpertDown, metal::MetalBuffer diagnostics,
    uint32_t rows, uint32_t selections) const {
  if (!gatheredQMVEnabled()) fail("private gathered I8 QMV down requires frozen flag1");
  const auto g = gathered_i8_qmv::geometry(rows, selections);
  const auto &layer = impl_->layer(index);
  if (impl_->metadata.layers[index].selectedIDs.size() != 512)
    fail("private gathered I8 QMV requires Full512");
  gatheredQMVViews(g, canonicalIntermediate, originalExpertIDs, canonicalExpertDown, diagnostics, true);
  for (const auto &buffer : {canonicalIntermediate, originalExpertIDs, canonicalExpertDown, diagnostics})
    impl_->immutableDisjoint(buffer);
  requireBytes(layer.ranks, 512 * 4);
  requireBytes(layer.codes[2], uint64_t{512} * 2560 * 640);
  requireBytes(layer.scales[2], uint64_t{512} * 2560 * 4);
  graph.add("flash_gathered_i8_qmv_down_sg4_c1", {canonicalIntermediate, layer.codes[2],
      layer.scales[2], layer.ranks, originalExpertIDs, canonicalExpertDown, diagnostics},
      FlashGatheredI8QMVParams{rows, selections, 512, 0},
      {g.downColumnGroups, rows, selections}, {gathered_i8_qmv::kThreads, 1, 1});
  impl_->recordGraph(true, rows, 512);
  impl_->qmvDownCalls.fetch_add(1, std::memory_order_relaxed);
  impl_->qmvDownRows.fetch_add(rows, std::memory_order_relaxed);
}
'''


def stage(source: Path, destination: Path) -> dict:
    if source.resolve() == destination.resolve():
        raise ValueError("private QMV staging requires a different destination")
    if destination.exists():
        raise ValueError("private QMV staging requires a NEW destination")
    destination.mkdir(parents=True)
    witness = {}
    for relative in ("runtime/flash/FlashInt8ExpertStore.hpp", "runtime/flash/FlashInt8ExpertStore.mm"):
        before = (source / relative).read_text()
        after = transform(relative, before)
        target = destination / relative
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(after)
        witness[relative] = {"input_sha256": hashlib.sha256(before.encode()).hexdigest(),
                             "output_sha256": hashlib.sha256(after.encode()).hexdigest()}
    for relative, content in extra_files().items():
        target = destination / relative
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(content)
        witness[relative] = {"output_sha256": hashlib.sha256(content.encode()).hexdigest()}
    result = {"policy": POLICY, "flag": FLAG, "gpu_work": False, "files": witness}
    (destination / "qmv-overlay-manifest.json").write_text(json.dumps(result, indent=2) + "\n")
    return result


def cpu_self_test(source: Path) -> dict:
    """Bounded source-only transform guards; no model metadata/payload reads."""
    outputs = {}
    for relative in ("runtime/flash/FlashInt8ExpertStore.hpp", "runtime/flash/FlashInt8ExpertStore.mm"):
        before = (source / relative).read_text()
        after = transform(relative, before)
        if after == before:
            raise AssertionError("supported source was not transformed")
        try:
            transform(relative, after)
        except RuntimeError:
            pass
        else:
            raise AssertionError("duplicate transform was not rejected")
        outputs[relative] = after
        for primitive in ("backend.allocateBuffer(", "backend.wrapSharedMemory(", "std::make_shared<Mapping>("):
            if after.count(primitive) != before.count(primitive):
                raise AssertionError("QMV added an allocation or mapping")
        if relative.endswith(".mm"):
            start = before.index("void FlashInt8ExpertStore::addGateUp(")
            end = before.index("} // namespace splash::flash", start)
            if before[start:end] not in after:
                raise AssertionError("original MPP graph methods changed")
            if "if (gatheredQMV)\n      derivative +=" not in after:
                raise AssertionError("conditional derivative suffix missing")
            if after.count("gathered_i8_qmv::requested()") != 2:
                raise AssertionError("frozen policy capture/recheck differs")
            if after.count("impl_->immutableDisjoint(buffer);") != before.count("impl_->immutableDisjoint(buffer);") + 2:
                raise AssertionError("all immutable layer disjoint checks missing")
    for relative in ("runtime/flash/FlashMTP.cpp", "runtime/flash/FlashForward.cpp", "runtime/flash/FlashBatchVerify.cpp"):
        if transform(relative, "untouched") != "untouched":
            raise AssertionError("Store transform rewrote executor/MTP source")
    # Independent identity construction proves flag0 control keeps the old hash.
    seed = "splash.private-allrows-target-v1\nsource=source\nstore=store\npolicy=mpp\nmtp=original-trained-bank\n"
    original = hashlib.sha256(seed.encode()).hexdigest()
    disabled = hashlib.sha256((seed + "").encode()).hexdigest()
    enabled = hashlib.sha256((seed + f"small_row_policy={POLICY}\n").encode()).hexdigest()
    if original != disabled or original == enabled:
        raise AssertionError("MPP control and QMV derivative identities collide")
    files = extra_files()
    if POLICY not in files[HEADER_RELATIVE]:
        raise AssertionError("Python and shared-header numerical policies differ")
    return {"cpu_source_checks": "passed", "gpu_work": False, "extra_allocations": 0,
            "numerical_qualification": "pending", "flag0_identity_preserved": True}


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--source", type=Path, required=True)
    parser.add_argument("--destination", type=Path)
    parser.add_argument("--cpu-self-test", action="store_true")
    args = parser.parse_args()
    if args.cpu_self_test:
        print(json.dumps(cpu_self_test(args.source), indent=2))
    else:
        if args.destination is None:
            parser.error("staging requires --destination")
        print(json.dumps(stage(args.source, args.destination), indent=2))
